File: type_checking/operations.py
from enum import Enum, auto

from data_types import *
from exceptions import *
from tokenization import BinaryOperator, UnaryOperator


class InvalidOperandType(CompilationException):
    def __init__(self, location: Location, operator: UnaryOperator, operand_type: DataType) -> None:
        super().__init__(location, f'Invalid operand type for {operator}: {operand_type}')


class UnaryOperationType(Enum):
    NEGATION = auto()
    BOOL_NORMALIZATION = auto()
    OPPOSITION = auto()


class UnaryOperation:
    @staticmethod
    def _invalid_operand_type_exception(location: Location, operator: UnaryOperator,
                                        operand_type: DataType) -> CompilationException:
        return CompilationException(location, f'Invalid operand type for {operator}: {operand_type}')

    @classmethod
    def from_operator(cls, location: Location, operator: UnaryOperator, operand_type: DataType) -> 'UnaryOperation':
        # !
        if operator is UnaryOperator.BANG:
            if operand_type == Types.CHAR:
                return cls(UnaryOperationType.NEGATION)
            raise InvalidOperandType(location, operator, operand_type)
        # !!
        if operator is UnaryOperator.DOUBLE_BANG:
            if operand_type == Types.CHAR:
                return cls(UnaryOperationType.BOOL_NORMALIZATION)
            raise InvalidOperandType(location, operator, operand_type)
        # -
        if operator is UnaryOperator.MINUS:
            if operand_type == Types.CHAR:
                return cls(UnaryOperationType.OPPOSITION)
            raise InvalidOperandType(location, operator, operand_type)
        raise CompilerException(f'Unknown unary operator: {operator}')

    def __init__(self, operation_type: UnaryOperationType, return_type: DataType = Types.CHAR) -> None:
        self.operation_type = operation_type
        self.return_type = return_type


class InvalidLeftOperandType(CompilationException):
    def __init__(self, location: Location, operator: BinaryOperator, operand_type: DataType) -> None:
        super().__init__(location, f'Invalid left operand type for {operator}: {operand_type}')


class InvalidRightOperandType(CompilationException):
    def __init__(self, location: Location, operator: BinaryOperator, operand_type: DataType) -> None:
        super().__init__(location, f'Invalid right operand type for {operator}: {operand_type}')


class NotEqualOperandTypes(CompilationException):
    def __init__(self, location: Location, operator: BinaryOperator, left_type: DataType, right_type: DataType) -> None:
        message = f'Invalid operand types for {operator}: both operand must be of the same type,' \
                  f' but found {left_type} and {right_type}'
        super().__init__(location, message)


class BinaryOperationType(Enum):
    EQUALITY_TEST = auto()
    DIFFERENCE_TEST = auto()
    STRICT_INEQUALITY_TEST = auto()
    LARGE_INEQUALITY_TEST = auto()
    INVERSE_STRICT_INEQUALITY_TEST = auto()
    INVERSE_LARGE_INEQUALITY_TEST = auto()
    CONJUNCTION = auto()
    DISJUNCTION = auto()
    ADDITION = auto()
    SUBTRACTION = auto()
    MULTIPLICATION = auto()
    DIVISION = auto()
    MODULO_OPERATION = auto()
    CONCATENATION = auto()


class BinaryOperation:
    @classmethod
    def from_operator(cls, location: Location, operator: BinaryOperator, left_type: DataType,
                      right_type: DataType) -> 'BinaryOperation':
        # ==
        if operator is BinaryOperator.DOUBLE_EQUAL:
            if left_type == right_type:
                return cls(BinaryOperationType.EQUALITY_TEST)
            raise NotEqualOperandTypes(location, operator, left_type, right_type)
        # !=
        if operator is BinaryOperator.BANG_EQUAL:
            if left_type == right_type:
                return cls(BinaryOperationType.DIFFERENCE_TEST)
            raise NotEqualOperandTypes(location, operator, left_type, right_type)
        # <
        if operator is BinaryOperator.LESS_THAN:
            if left_type == Types.CHAR:
                if right_type == Types.CHAR:
                    return cls(BinaryOperationType.STRICT_INEQUALITY_TEST)
                raise InvalidRightOperandType(location, operator, right_type)
            raise InvalidLeftOperandType(location, operator, left_type)
        # <=
        if operator is BinaryOperator.LESS_THAN_EQUAL:
            if left_type == Types.CHAR:
                if right_type == Types.CHAR:
                    return cls(BinaryOperationType.LARGE_INEQUALITY_TEST)
                raise InvalidRightOperandType(location, operator, right_type)
            raise InvalidLeftOperandType(location, operator, left_type)
        # >
        if operator is BinaryOperator.GREATER_THAN:
            if left_type == Types.CHAR:
                if right_type == Types.CHAR:
                    return cls(BinaryOperationType.INVERSE_STRICT_INEQUALITY_TEST)
                raise InvalidRightOperandType(location, operator, right_type)
            raise InvalidLeftOperandType(location, operator, left_type)
        # >=
        if operator is BinaryOperator.GREATER_THAN_EQUAL:
            if left_type == Types.CHAR:
                if right_type == Types.CHAR:
                    return cls(BinaryOperationType.INVERSE_LARGE_INEQUALITY_TEST)
                raise InvalidRightOperandType(location, operator, right_type)
            raise InvalidLeftOperandType(location, operator, left_type)
        # &&
        if operator is BinaryOperator.DOUBLE_AMPERSAND:
            if left_type == Types.CHAR:
                if right_type == Types.CHAR:
                    return cls(BinaryOperationType.CONJUNCTION)
                raise InvalidRightOperandType(location, operator, right_type)
            raise InvalidLeftOperandType(location, operator, left_type)
        # ||
        if operator is BinaryOperator.DOUBLE_PIPE:
            if left_type == Types.CHAR:
                if right_type == Types.CHAR:
                    return cls(BinaryOperationType.DISJUNCTION)
                raise InvalidRightOperandType(location, operator, right_type)
            raise InvalidLeftOperandType(location, operator, left_type)
        # +
        if operator is BinaryOperator.PLUS:
            if left_type == Types.CHAR:
                if right_type == Types.CHAR:
                    return cls(BinaryOperationType.ADDITION)
                raise InvalidRightOperandType(location, operator, right_type)
            raise InvalidLeftOperandType(location, operator, left_type)
        # -
        if operator is BinaryOperator.MINUS:
            if left_type == Types.CHAR:
                if right_type == Types.CHAR:
                    return cls(BinaryOperationType.SUBTRACTION)
                raise InvalidRightOperandType(location, operator, right_type)
            raise InvalidLeftOperandType(location, operator, left_type)
        # *
        if operator is BinaryOperator.STAR:
            if left_type == Types.CHAR:
                if right_type == Types.CHAR:
                    return cls(BinaryOperationType.MULTIPLICATION)
                raise InvalidRightOperandType(location, operator, right_type)
            raise InvalidLeftOperandType(location, operator, left_type)
        # /
        if operator is BinaryOperator.SLASH:
            if left_type == Types.CHAR:
                if right_type == Types.CHAR:
                    return cls(BinaryOperationType.DIVISION)
                raise InvalidRightOperandType(location, operator, right_type)
            raise InvalidLeftOperandType(location, operator, left_type)
        # %
        if operator is BinaryOperator.PERCENT:
            if left_type == Types.CHAR:
                if right_type == Types.CHAR:
                    return cls(BinaryOperationType.MODULO_OPERATION)
                raise InvalidRightOperandType(location, operator, right_type)
            raise InvalidLeftOperandType(location, operator, left_type)
        # ..
        if operator is BinaryOperator.DOUBLE_DOT:
            if isinstance(left_type, ArrayType):
                if isinstance(right_type, ArrayType):
                    if left_type.base_type == right_type.base_type:
                        total_size = left_type.size() + right_type.size()
                        return cls(BinaryOperationType.CONCATENATION, ArrayType(left_type.base_type, total_size))
                raise InvalidRightOperandType(location, operator, right_type)
            raise InvalidLeftOperandType(location, operator, left_type)
        raise CompilerException(f'Unknown binary operator: {operator}')

    def __init__(self, operation_type: BinaryOperationType, return_type: DataType = Types.CHAR) -> None:
        self.operation_type = operation_type
        self.return_type = return_type


__all__ = ['UnaryOperationType', 'UnaryOperation', 'BinaryOperationType', 'BinaryOperation']

File: src/type_checker/operations.rs
use std::fmt::{Display, Formatter};
use std::rc::Rc;

use crate::exceptions::{CompilationException, CompilationResult};
use crate::lexer::lexemes::Symbol;
use crate::location::{Located, Location};
use crate::type_checker::types::{Type, Value};
use crate::utils::product::Product;

/// An `Operation` has a [result type](Self::result_type) and can be [applied to values](Self::apply).
pub trait Operation<const N: usize>: Copy {
    /// Returns the operation corresponding to an operator and the types of its operands.
    fn from_untyped(location: Location, operator: Symbol, operands: [&Type; N]) -> CompilationResult<Self>;

    /// Returns the type of the result of this operation.
    fn result_type(self) -> Type;

    /// Applies this operation to a constant value.
    fn apply(self, operands: [Located<Value>; N]) -> CompilationResult<Value>;
}


#[derive(Copy, Clone, Debug)]
pub enum UnaryOperation {
    Negation,
    BooleanNormalization,
    Opposition,
}

impl Operation<1> for UnaryOperation {
    fn from_untyped(location: Location, operator: Symbol, operands: [&Type; 1]) -> CompilationResult<Self> {
        match (operator, operands) {
            (Symbol::Bang, [Type::Char]) => Ok(Self::Negation),
            (Symbol::DoubleBang, [Type::Char]) => Ok(Self::BooleanNormalization),
            (Symbol::Minus, [Type::Char]) => Ok(Self::Opposition),
            _ => Err(CompilationException::invalid_operator(location, operator, &operands))
        }
    }

    fn result_type(self) -> Type {
        match self {
            Self::Negation => Type::Char,
            Self::BooleanNormalization => Type::Char,
            Self::Opposition => Type::Char,
        }
    }

    fn apply(self, [operand]: [Located<Value>; 1]) -> CompilationResult<Value> {
        match self {
            Self::Negation => Ok(Value::bool(!operand.into_bool()?)),
            Self::BooleanNormalization => Ok(Value::bool(operand.into_bool()?)),
            Self::Opposition => Ok(Value::Char(operand.into_char()?.wrapping_neg())),
        }
    }
}

impl Display for UnaryOperation {
    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
        match self {
            Self::Negation => write!(f, "negation"),
            Self::BooleanNormalization => write!(f, "boolean normalization"),
            Self::Opposition => write!(f, "opposition"),
        }
    }
}


#[derive(Copy, Clone, Debug)]
pub enum BinaryOperation {
    /// Tests if two values of the same type are equal.
    EqualityTest,
    /// Tests if two values of the same type are different.
    DifferenceTest,
    /// Tests if the left operand is strictly less than the right operand.
    StrictInequalityTest,
    /// Tests if the left operand is less than or equal to the right operand.
    LargeInequalityTest,
    /// Tests if the left operand is strictly greater than the right operand.
    InverseStrictInequalityTest,
    /// Tests if the left operand is greater than or equal to the right operand.
    InverseLargeInequalityTest,
    /// Computes the logical conjunction of two [`char`s](Type::Char).
    Conjunction,
    /// Computes the logical disjunction of two [`char`s](Type::Char).
    ///
    /// The result of this operation is a boolean value (either 0 or 1). To get the value of the
    /// first non-zero operand, use [`Self::Disjunction`].
    BooleanDisjunction,
    /// Returns the left-hand side if non-zero, the right-hand side otherwise.
    ///
    /// To get a normalized boolean value (either 0 or 1), use [`Self::BooleanDisjunction`].
    Disjunction,
    /// Computes the sum of two [`char`s](Type::Char).
    ///
    /// Overflows result in a wrap around.
    Addition,
    /// Computes the subtraction of two [`char`s](Type::Char).
    ///
    /// Overflows result in a wrap around.
    Subtraction,
    /// Computes the product of two [`char`s](Type::Char).
    ///
    /// Overflows result in a wrap around.
    Multiplication,
    /// Computes the floor division of two [`char`s](Type::Char).
    ///
    /// Division by zero has an undefined result.
    Division,
    /// Computes the remainder of the euclidean division of two [`char`s](Type::Char).
    ///
    /// Modulo zero has an undefined result.
    Modulo,
    /// Computes the euclidean division of two [`char`s](Type::Char).
    ///
    /// Division by zero has an undefined result.
    EuclideanDivision,
}

fn division(lhs: u8, rhs: u8) -> u8 {
    if rhs == 0 {
        0
    } else {
        lhs.wrapping_div(rhs)
    }
}

fn modulo(lhs: u8, rhs: u8) -> u8 {
    if rhs == 0 {
        0
    } else {
        lhs.wrapping_rem(rhs)
    }
}

impl Operation<2> for BinaryOperation {
    fn from_untyped(location: Location, operator: Symbol, operands: [&Type; 2]) -> CompilationResult<Self> {
        match (operator, operands) {
            (Symbol::DoubleEqual, _) => Ok(Self::EqualityTest),
            (Symbol::BangEqual, _) => Ok(Self::DifferenceTest),
            (Symbol::LessThan, [Type::Char, Type::Char]) => Ok(Self::StrictInequalityTest),
            (Symbol::LessThanEqual, [Type::Char, Type::Char]) => Ok(Self::LargeInequalityTest),
            (Symbol::GreaterThan, [Type::Char, Type::Char]) => Ok(Self::InverseStrictInequalityTest),
            (Symbol::GreaterThanEqual, [Type::Char, Type::Char]) => Ok(Self::InverseLargeInequalityTest),
            (Symbol::DoubleAmpersand, [Type::Char, Type::Char]) => Ok(Self::Conjunction),
            (Symbol::DoublePipe, [Type::Char, Type::Char]) => Ok(Self::BooleanDisjunction),
            (Symbol::DoubleQuestionMark, [Type::Char, Type::Char]) => Ok(Self::Disjunction),
            (Symbol::Plus, [Type::Char, Type::Char]) => Ok(Self::Addition),
            (Symbol::Minus, [Type::Char, Type::Char]) => Ok(Self::Subtraction),
            (Symbol::Star, [Type::Char, Type::Char]) => Ok(Self::Multiplication),
            (Symbol::Slash, [Type::Char, Type::Char]) => Ok(Self::Division),
            (Symbol::Percent, [Type::Char, Type::Char]) => Ok(Self::Modulo),
            (Symbol::DoublePercent, [Type::Char, Type::Char]) => Ok(Self::EuclideanDivision),
            _ => Err(CompilationException::invalid_operator(location, operator, &operands))
        }
    }

    fn result_type(self) -> Type {
        match self {
            Self::EqualityTest => Type::Char,
            Self::DifferenceTest => Type::Char,
            Self::StrictInequalityTest => Type::Char,
            Self::LargeInequalityTest => Type::Char,
            Self::InverseStrictInequalityTest => Type::Char,
            Self::InverseLargeInequalityTest => Type::Char,
            Self::Conjunction => Type::Char,
            Self::BooleanDisjunction => Type::Char,
            Self::Disjunction => Type::Char,
            Self::Addition => Type::Char,
            Self::Subtraction => Type::Char,
            Self::Multiplication => Type::Char,
            Self::Division => Type::Char,
            Self::Modulo => Type::Char,
            Self::EuclideanDivision => Type::from_operands(&[Type::Char, Type::Char]),
        }
    }

    fn apply(self, [lhs, rhs]: [Located<Value>; 2]) -> CompilationResult<Value> {
        match self {
            BinaryOperation::EqualityTest => Ok(Value::bool(lhs == rhs)),
            BinaryOperation::DifferenceTest => Ok(Value::bool(lhs != rhs)),
            BinaryOperation::StrictInequalityTest => Ok(Value::bool(lhs.into_char()? < rhs.into_char()?)),
            BinaryOperation::LargeInequalityTest => Ok(Value::bool(lhs.into_char()? <= rhs.into_char()?)),
            BinaryOperation::InverseStrictInequalityTest => Ok(Value::bool(lhs.into_char()? > rhs.into_char()?)),
            BinaryOperation::InverseLargeInequalityTest => Ok(Value::bool(lhs.into_char()? >= rhs.into_char()?)),
            BinaryOperation::Conjunction => Ok(Value::bool(lhs.into_bool()? && rhs.into_bool()?)),
            BinaryOperation::BooleanDisjunction => Ok(Value::bool(lhs.into_bool()? || rhs.into_bool()?)),
            BinaryOperation::Disjunction => Ok(Value::Char({
                let l = lhs.into_char()?;
                let r = rhs.into_char()?;
                if l == 0 { r } else { l }
            })),
            BinaryOperation::Addition => Ok(Value::Char(lhs.into_char()?.wrapping_add(rhs.into_char()?))),
            BinaryOperation::Subtraction => Ok(Value::Char(lhs.into_char()?.wrapping_sub(rhs.into_char()?))),
            BinaryOperation::Multiplication => Ok(Value::Char(lhs.into_char()?.wrapping_mul(rhs.into_char()?))),
            BinaryOperation::Division => Ok(Value::Char(division(lhs.into_char()?, rhs.into_char()?))),
            BinaryOperation::Modulo => Ok(Value::Char(modulo(lhs.into_char()?, rhs.into_char()?))),
            BinaryOperation::EuclideanDivision => Ok({
                let l = lhs.into_char()?;
                let r = rhs.into_char()?;
                let div = Value::Char(division(l, r));
                let rem = Value::Char(modulo(l, r));
                Value::Tuple(Rc::new(Product::new_minimal([div, rem])))
            }),
        }
    }
}

impl Display for BinaryOperation {
    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
        match self {
            Self::EqualityTest => write!(f, "equality test"),
            Self::DifferenceTest => write!(f, "difference test"),
            Self::StrictInequalityTest => write!(f, "strict inequality test"),
            Self::LargeInequalityTest => write!(f, "large inequality test"),
            Self::InverseStrictInequalityTest => write!(f, "inverse strict equality test"),
            Self::InverseLargeInequalityTest => write!(f, "inverse large inequality test"),
            Self::Conjunction => write!(f, "conjunction"),
            Self::BooleanDisjunction => write!(f, "boolean disjunction"),
            Self::Disjunction => write!(f, "disjunction"),
            Self::Addition => write!(f, "addition"),
            Self::Subtraction => write!(f, "subtraction"),
            Self::Multiplication => write!(f, "multiplication"),
            Self::Division => write!(f, "division"),
            Self::Modulo => write!(f, "modulo"),
            Self::EuclideanDivision => write!(f, "euclidean division"),
        }
    }
}
